calculator/calculator.py

```diff
@@ -3,109 +3,91 @@
-
-
-def replace_in_expression(operator_regex: str, expression: str,
-                          new_expr: str) -> str:
-    """Change the old sub expression to a new one according to pattern.
+OPERATORS = [
+    Operator(precedence=1, operation=operator.add,
+             regex=fr"({ANY_NUMBER})\+({ANY_NUMBER})"),
+    Operator(precedence=1, operation=operator.sub,
+             regex=fr"({ANY_NUMBER})-({ANY_NUMBER})"),
+    Operator(precedence=2, operation=operator.mul,
+             regex=fr"({ANY_NUMBER})\*({ANY_NUMBER})"),
+    Operator(precedence=2, operation=operator.truediv,
+             regex=fr"({ANY_NUMBER})/({ANY_NUMBER})"),
+    Operator(precedence=3, operation=math.pow,
+             regex=fr"({ANY_NUMBER})\^({ANY_NUMBER})"),
+    Operator(precedence=4, operation=math.fmod,
+             regex=fr"({ANY_NUMBER})%({ANY_NUMBER})"),
+    Operator(precedence=5, operation=lambda x, y: (x + y) / 2.0,
+             regex=fr"({ANY_NUMBER})@({ANY_NUMBER})"),
+    Operator(precedence=5, operation=max,
+             regex=fr"({ANY_NUMBER})\$({ANY_NUMBER})"),
+    Operator(precedence=5, operation=min,
+             regex=fr"({ANY_NUMBER})&({ANY_NUMBER})"),
+    Operator(precedence=6, operation=operator.neg,
+             regex=fr"~({ANY_NUMBER})"),
+    Operator(precedence=7, operation=math.factorial,
+             regex=fr"({ANY_NUMBER})!"),
+    Operator(precedence=maxsize, operation=lambda expr: evaluate(expr),
+             regex=r"\(([^()]*)\)")
+]
+
+
+def evaluate(expression: str) -> float:
+    """Calculate the value of a mathematical expression.
 
     Args:
-        operator_regex: pattern to find a sub expression of the operator.
-        expression: the entire expression.
-        new_expr: the expression to put instead of the old one.
+        expression: the entire mathematical expression to calculate.
 
     Returns:
-        A new whole expression, changed according to pattern.
+         The calculated result of the given expression.
     """
-    old_expression = re.search(operator_regex, expression).group()
-    return expression.replace(old_expression, new_expr)
-
-
-def _evaluate(expression: str) -> float:
-    """Calculate the value of a mathematical expression."""
-    expression = str(re.sub(r"\s+", "", expression))
-    operators_by_precedence = operators_info_by_precedence()
+    expression = expression.replace(" ", "")
+    operators_by_precedence = sorted(OPERATORS, reverse=True,
+                                     key=operator.attrgetter("precedence"))
 
     for operator_info in operators_by_precedence:
         operator_regex = operator_info.regex
 
-        try:
-            operands = list(re.search(operator_regex, expression).groups())
+        search_result = re.search(operator_regex, expression)
+
+        if search_result is not None:
+            operands = search_result.groups()
+
+            if all(True if re.match(f"^{ANY_NUMBER}$", op) else False for op
+                   in operands):
```

Comment:
You dont need the True False part, the value itself will be treated as a true or false.